In [220]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [221]:
def non_ascii_words(n):
    ASCII = '\x00\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x0b\x0c\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\x7f'
    for ch in n.upper():
        if ch not in ASCII:
            return True
    return False

In [222]:
survey=pd.read_csv("data_640_validated-V1.0.csv")
#print(survey.head(20))
questions = pd.read_csv("Data description_validated.csv")
#print(questions.head(20))

In [223]:
#l = survey.items()
for i,values in survey.items():
  print(f'{i} - {survey[i].unique()}', end='\n\n')

Gamer_ID - [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180
 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198
 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216
 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234
 235 236 237 238 239 240 241 242 243 244

In [224]:
import math 

for li,values in survey.items():
    print(li + " : ")
    count = 2
    totcount = 0
    for i,row in survey.iterrows():
        if type(row[li]) == float:
            if math.isnan(row[li]):
                if li=="A4":
                    survey.loc[i,li] = "None"
                    print(count)
                if li=="D2":
                    survey.loc[i,li] = "Other"
                    print(count)
                if li=="D7":
                    survey.loc[i,li] = "Other"
                    print(count)
                totcount = totcount + 1
        count = count + 1
    print("TOTCOUNT : " + str(totcount))
        #print('\n')

Gamer_ID : 
TOTCOUNT : 0
Date of Entry : 
TOTCOUNT : 0
A1_1 : 
TOTCOUNT : 0
Grouped A1_1 : 
TOTCOUNT : 0
A1_2 : 
TOTCOUNT : 0
A2 : 
TOTCOUNT : 0
A3 : 
TOTCOUNT : 0
A4 : 
2
9
10
11
13
15
17
20
21
22
24
25
30
33
43
51
66
71
72
93
102
135
136
159
201
211
213
218
243
246
249
252
255
271
272
284
294
296
345
362
365
366
377
382
390
395
402
419
420
421
434
441
442
452
453
454
458
467
468
472
474
475
478
481
490
495
507
516
517
528
537
538
551
555
560
569
574
579
588
589
590
595
598
609
610
641
TOTCOUNT : 86
A5 : 
TOTCOUNT : 0
A6 : 
TOTCOUNT : 0
A7 : 
TOTCOUNT : 0
A8 : 
TOTCOUNT : 0
B1 : 
TOTCOUNT : 0
B2 : 
TOTCOUNT : 0
B3 : 
TOTCOUNT : 0
C1 : 
TOTCOUNT : 0
C2 : 
TOTCOUNT : 0
C3 : 
TOTCOUNT : 0
C4 : 
TOTCOUNT : 0
C5 : 
TOTCOUNT : 0
C6 : 
TOTCOUNT : 0
C7 : 
TOTCOUNT : 0
C8 : 
TOTCOUNT : 0
C9 : 
TOTCOUNT : 0
C10 : 
TOTCOUNT : 0
C11 : 
TOTCOUNT : 0
C12 : 
TOTCOUNT : 0
C13 : 
TOTCOUNT : 0
C14 : 
TOTCOUNT : 0
C15 : 
TOTCOUNT : 0
D1 : 
TOTCOUNT : 0
D2 : 
61
82
83
93
189
TOTCOUNT : 5
D3 : 
TOTCOUNT :

In [225]:
l = []
for j,values in survey.items():
    if values.isnull().sum()!= 0:
        print(j + " : ")
        print(values.isnull().sum())
        l.append(j)


In [226]:
import math 

for li,values in survey.items():
    print(li + " : ")
    count = 2
    totcount = 0
    for i,row in survey.iterrows():
        if type(row[li]) == str:
            if non_ascii_words(row[li]) or row[li]=="" or row[li]==" ":
                print(count)
                #print(row[li])
                totcount = totcount + 1
        count = count + 1
    print("TOTCOUNT : " + str(totcount))
        #print('\n')

Gamer_ID : 
TOTCOUNT : 0
Date of Entry : 
TOTCOUNT : 0
A1_1 : 
TOTCOUNT : 0
Grouped A1_1 : 
TOTCOUNT : 0
A1_2 : 
TOTCOUNT : 0
A2 : 
TOTCOUNT : 0
A3 : 
TOTCOUNT : 0
A4 : 
TOTCOUNT : 0
A5 : 
TOTCOUNT : 0
A6 : 
TOTCOUNT : 0
A7 : 
TOTCOUNT : 0
A8 : 
TOTCOUNT : 0
B1 : 
TOTCOUNT : 0
B2 : 
TOTCOUNT : 0
B3 : 
TOTCOUNT : 0
C1 : 
TOTCOUNT : 0
C2 : 
TOTCOUNT : 0
C3 : 
TOTCOUNT : 0
C4 : 
TOTCOUNT : 0
C5 : 
TOTCOUNT : 0
C6 : 
TOTCOUNT : 0
C7 : 
TOTCOUNT : 0
C8 : 
TOTCOUNT : 0
C9 : 
TOTCOUNT : 0
C10 : 
TOTCOUNT : 0
C11 : 
TOTCOUNT : 0
C12 : 
TOTCOUNT : 0
C13 : 
TOTCOUNT : 0
C14 : 
TOTCOUNT : 0
C15 : 
TOTCOUNT : 0
D1 : 
TOTCOUNT : 0
D2 : 
384
TOTCOUNT : 1
D3 : 
TOTCOUNT : 0
D4 : 
273
377
381
TOTCOUNT : 3
D5 : 
TOTCOUNT : 0
D6 : 
133
297
305
TOTCOUNT : 3
D7 : 
43
57
101
105
108
132
215
297
313
337
383
397
422
438
439
454
460
474
494
506
509
526
527
546
589
593
611
TOTCOUNT : 27
E1 : 
TOTCOUNT : 0
E2 : 
TOTCOUNT : 0
E3 : 
TOTCOUNT : 0
E4 : 
TOTCOUNT : 0
E5 : 
TOTCOUNT : 0
E6 : 
TOTCOUNT : 0
E7 : 
TOTCO

for i,row in questions.iterrows():
    if row["Code"] in l:
        print(row)

import math 

for li,values in survey.items():
    print(li + " : ")
    count = 2
    totcount = 0
    for i,row in survey.iterrows():
        if row[li] is None:
            print(count)
            totcount = totcount + 1
        count = count + 1
    print("TOTCOUNT : " + str(totcount))
        #print('\n')

1. There are 86 nan values in column A4
2. There are 6 nan values in column D1 
3. There are 5 nan values and 1 non ascii string in column D2
4. There are 1 nan values in column D3
5. There are 1 nan values and 3 non ascii strings in column D4
6. There are 3 non ascii strings in column D6
7. There are 13 nan values and 27 non ascii in column D7 

import math 

for li,values in survey.items():
    print(li + " : ")
    count = 2
    totcount = 0
    for i,row in survey.iterrows():
        if type(row[li]) != int:
            if row[li].isna():
                print(count)
                totcount = totcount + 1
        count = count + 1
    print("TOTCOUNT : " + str(totcount))
        #print('\n')

survey.loc[0,"A4"]

for li,values in survey.items():
    #print(li + " : ")
    count = 2
    totcount = 0
    for i,row in survey.iterrows():
        if type(row[li]) == str:
            if non_ascii_words(row[li]) or row[li]=="" or row[li]==" ":
                #print(count)
                survey.loc[i,li] = "DNC"
                #print(row[li])
                totcount = totcount + 1
                #col = li
                #rah = i
    #print("TOTCOUNT : " + str(totcount))
        #print('\n')
#print(survey.loc[rah , col])

survey = survey.replace(np.nan,"DNC")

count = 1
for i,values in survey.items():
    if count >= 3:
        print("Question : "+questions.iloc[count-3]["Question"],end='\n' )
        print("Group : "+questions.iloc[count-3]["Group"],end='\n' )
    
    print(f'{i} - {survey[i].unique()}', end='\n\n')
    count = count + 1

survey.loc[0,"A4"]

survey.loc[382,"D2"]

In [227]:
def isSubstring(s1, s2):
    M = len(s1)
    N = len(s2)
 
    # A loop to slide pat[] one by one
    for i in range(N - M + 1):
 
        # For current index i,
        # check for pattern match
        for j in range(M):
            if (s2[i + j] != s1[j]):
                break
 
        if j + 1 == M:
            return i
 
    return -1

In [228]:
s = survey["D2"].unique()
#print(s)
words = []
wordss = []
for ss in s:
    words.append(''.join(ss.split()))#.split("\n")).split(" ")
words = ''.join(words)
#print(words)
#for word in words:
#    wordss.append(word.split(","))
wordss = words.split(",")
wordss = set(wordss)
wordss = sorted(list(wordss))
for i in range(0,len(wordss)):
    for j in range(0,len(wordss)):
        if i!=j and wordss[j]!="Remove":
        
            if isSubstring(wordss[i].upper(),wordss[j].upper())!= -1:
                if len(wordss[j])>len(wordss[i]):
                    wordss[j] = "Remove"
                else:
                    wordss[i] = "Remove"
#wordss.remove("Remove")
words = []
for word in wordss:
    if word!="Remove":
        words.append(word)
        
words
        
#wordss = list(set(wordss))
#wordss.remove(" ")
#wordss.remove(",")
#words = set(words)
#words = list(words)
#wordss

['Action',
 'Adventure',
 'FPS',
 'Fighting',
 'Pointandclick',
 'RPG',
 'Simulation',
 'Sports',
 'Story-driven',
 'Strategy',
 'Variedthemesallinonegameex:recroom',
 'puzzle',
 'vrchat']

In [229]:

arr = [0]*len(survey.index)
for w in words:
    survey.assign(w = arr)
    for i,rows in survey.iterrows(): 
        if isSubstring(w.upper(),survey.loc[i,"D2"].upper())!= -1:
            survey.loc[i,w] = "Yes"
        else:
            survey.loc[i,w] = "No"
survey

Gamer_ID    Date of Entry        A1_1 Grouped A1_1       A1_2      A2  \
0           1  5/15/2020 14:20         USA     American  US/Canada    Male   
1           2  5/15/2020 18:37     Vietnam   Vietnamese       Asia  Female   
2           3   5/16/2020 0:21     Vietnam   Vietnamese       Asia  Female   
3           4   5/16/2020 0:25     Vietnam   Vietnamese       Asia  Female   
4           5   5/16/2020 0:26     Vietnam   Vietnamese       Asia    Male   
..        ...              ...         ...          ...        ...     ...   
635       636    06/08/20 9:22         usa     American  US/Canada    Male   
636       637   06/08/20 16:29         usa     American  US/Canada    Male   
637       638   06/08/20 21:42          uk      British         EU    Male   
638       639    06/09/20 0:36          uk      British         EU    Male   
639       640  6/15/2020 18:51  Vietnamese   Vietnamese       Asia  Female   

                             A3        A4  A5                         A6  ...  \
0          Undergraduate school      None  22                      Asian  ...   
1          Undergraduate school  A garden  23                      Asian  ...   
2                   High school  A garden  23                      Asian  ...   
3    Graduate school and higher     A pet  31                      Asian  ...   
4    Graduate school and higher     A pet  21                      Asian  ...   
..                          ...       ...  ..                        ...  ...   
635        Undergraduate school      Both  28                      White  ...   
636  Graduate school and higher      Both  27  Black or African American  ...   
637        Undergraduate school      Both  30  Black or African American  ...   
638  Graduate school and higher      Both  31                      White  ...   
639        Undergraduate school      None  27                      Asian  ...   

    Fighting Pointandclick  RPG Simulation Sports  Story-driven  Strategy  \
0         No            No  Yes        Yes     No            No       Yes   
1         No            No   No        Yes     No            No       Yes   
2         No            No  Yes         No     No            No       Yes   
3         No            No  Yes        Yes     No            No       Yes   
4         No            No  Yes        Yes     No            No       Yes   
..       ...           ...  ...        ...    ...           ...       ...   
635       No            No  Yes        Yes    Yes            No       Yes   
636       No            No  Yes        Yes    Yes            No       Yes   
637       No            No  Yes        Yes    Yes            No       Yes   
638       No            No  Yes        Yes    Yes            No       Yes   
639       No            No  Yes         No     No            No       Yes   

     Variedthemesallinonegameex:recroom  puzzle  vrchat  
0                                    No      No      No  
1                                    No     Yes      No  
2                                    No      No      No  
3                                    No      No      No  
4                                    No      No      No  
..                                  ...     ...     ...  
635                                  No      No      No  
636                                  No      No      No  
637                                  No      No      No  
638                                  No      No      No  
639                                  No      No      No  

[640 rows x 110 columns]

s = survey["D6"].unique()
#print(s)
strings = ""
words = []
wordss = []
for ss in s:
    words.append(' '.join(ss.split()))#.split("\n")).split(" ")
words = ' '.join(words)
#print(words)
#for word in words:
#    wordss.append(word.split(","))
wordss = words.split(" ")
wordss = set(wordss)
wordss = sorted(list(wordss))
#print(wordss)
for i in range(0,len(wordss)):
    for j in range(0,len(wordss)):
        if len(wordss[j])<=3:
            wordss[j]="Remove"
        if i!=j and wordss[j]!="Remove":
        
            if isSubstring(wordss[i].upper(),wordss[j].upper())!= -1:
                if len(wordss[j])>len(wordss[i]):
                    wordss[j] = "Remove"
                else:
                    wordss[i] = "Remove"
#wordss.remove("Remove")
words = []
for word in wordss:
    if word!="Remove":
        words.append(word)
        
words

In [230]:
import re
for i, rows in survey.iterrows():
    if survey.loc[i,"D5"]=="Not yet play":
        survey.loc[i,"D5"] = "No"
    if survey.loc[i,"D5"]=="Once in a while. Dont play as much anymore." or survey.loc[i,"D5"]=="More often than I exist in reality":
        survey.loc[i,"D5"] = "1. Everyday"
    l = re.findall(r'\d+', str(survey.loc[i,"D4"]))
    SUM = 0
    for ele in l:
        SUM = SUM + int(ele)
    if SUM != 0:
        survey.loc[i,"D4"] = SUM/len(l)
    else:
        survey.loc[i,"D4"] = 3.5
survey

Gamer_ID    Date of Entry        A1_1 Grouped A1_1       A1_2      A2  \
0           1  5/15/2020 14:20         USA     American  US/Canada    Male   
1           2  5/15/2020 18:37     Vietnam   Vietnamese       Asia  Female   
2           3   5/16/2020 0:21     Vietnam   Vietnamese       Asia  Female   
3           4   5/16/2020 0:25     Vietnam   Vietnamese       Asia  Female   
4           5   5/16/2020 0:26     Vietnam   Vietnamese       Asia    Male   
..        ...              ...         ...          ...        ...     ...   
635       636    06/08/20 9:22         usa     American  US/Canada    Male   
636       637   06/08/20 16:29         usa     American  US/Canada    Male   
637       638   06/08/20 21:42          uk      British         EU    Male   
638       639    06/09/20 0:36          uk      British         EU    Male   
639       640  6/15/2020 18:51  Vietnamese   Vietnamese       Asia  Female   

                             A3        A4  A5                         A6  ...  \
0          Undergraduate school      None  22                      Asian  ...   
1          Undergraduate school  A garden  23                      Asian  ...   
2                   High school  A garden  23                      Asian  ...   
3    Graduate school and higher     A pet  31                      Asian  ...   
4    Graduate school and higher     A pet  21                      Asian  ...   
..                          ...       ...  ..                        ...  ...   
635        Undergraduate school      Both  28                      White  ...   
636  Graduate school and higher      Both  27  Black or African American  ...   
637        Undergraduate school      Both  30  Black or African American  ...   
638  Graduate school and higher      Both  31                      White  ...   
639        Undergraduate school      None  27                      Asian  ...   

    Fighting Pointandclick  RPG Simulation Sports  Story-driven  Strategy  \
0         No            No  Yes        Yes     No            No       Yes   
1         No            No   No        Yes     No            No       Yes   
2         No            No  Yes         No     No            No       Yes   
3         No            No  Yes        Yes     No            No       Yes   
4         No            No  Yes        Yes     No            No       Yes   
..       ...           ...  ...        ...    ...           ...       ...   
635       No            No  Yes        Yes    Yes            No       Yes   
636       No            No  Yes        Yes    Yes            No       Yes   
637       No            No  Yes        Yes    Yes            No       Yes   
638       No            No  Yes        Yes    Yes            No       Yes   
639       No            No  Yes         No     No            No       Yes   

     Variedthemesallinonegameex:recroom  puzzle  vrchat  
0                                    No      No      No  
1                                    No     Yes      No  
2                                    No      No      No  
3                                    No      No      No  
4                                    No      No      No  
..                                  ...     ...     ...  
635                                  No      No      No  
636                                  No      No      No  
637                                  No      No      No  
638                                  No      No      No  
639                                  No      No      No  

[640 rows x 110 columns]

In [231]:
survey.to_csv("THE_ULTIMATE_CLEAN_DATASET.csv")

In [232]:
survey.head(20)

Gamer_ID    Date of Entry        A1_1 Grouped A1_1       A1_2      A2  \
0          1  5/15/2020 14:20         USA     American  US/Canada    Male   
1          2  5/15/2020 18:37     Vietnam   Vietnamese       Asia  Female   
2          3   5/16/2020 0:21     Vietnam   Vietnamese       Asia  Female   
3          4   5/16/2020 0:25     Vietnam   Vietnamese       Asia  Female   
4          5   5/16/2020 0:26     Vietnam   Vietnamese       Asia    Male   
5          6   5/16/2020 0:29     Vietnam   Vietnamese       Asia  Female   
6          7   5/16/2020 0:30     Vietnam   Vietnamese       Asia    Male   
7          8   5/16/2020 0:34     Vietnam   Vietnamese       Asia  Female   
8          9   5/16/2020 0:36     Vietnam   Vietnamese       Asia    Male   
9         10   5/16/2020 0:42     Vietnam   Vietnamese       Asia  Female   
10        11   5/16/2020 0:50     Vietnam   Vietnamese       Asia    Male   
11        12   5/16/2020 0:52     Vietnam   Vietnamese       Asia    Male   
12        13   5/16/2020 0:55       China      Chinese       Asia  Female   
13        14   5/16/2020 1:08     Vietnam   Vietnamese       Asia    Male   
14        15   5/16/2020 1:16    Viet Nam   Vietnamese       Asia    Male   
15        16   5/16/2020 1:37     Vietnam   Vietnamese       Asia    Male   
16        17   5/16/2020 1:56     Vietnam   Vietnamese       Asia    Male   
17        18   5/16/2020 2:02     Vietnam   Vietnamese       Asia    Male   
18        19   5/16/2020 2:37  Vietnamese   Vietnamese       Asia    Male   
19        20   5/16/2020 2:54  Vietnamese   Vietnamese       Asia    Male   

                            A3        A4  A5     A6  ... Fighting  \
0         Undergraduate school      None  22  Asian  ...       No   
1         Undergraduate school  A garden  23  Asian  ...       No   
2                  High school  A garden  23  Asian  ...       No   
3   Graduate school and higher     A pet  31  Asian  ...       No   
4   Graduate school and higher     A pet  21  Asian  ...       No   
5                  High school     A pet  27  Asian  ...       No   
6   Graduate school and higher  A garden  26  Asian  ...       No   
7                  High school      None  26  Asian  ...       No   
8         Undergraduate school      None  20  Asian  ...       No   
9         Undergraduate school      None  20  Asian  ...       No   
10                 High school  A garden  17  Asian  ...       No   
11                 High school      None  23  Asian  ...       No   
12        Undergraduate school  A garden  22  Asian  ...       No   
13                 High school      None  20  Asian  ...       No   
14  Graduate school and higher      Both  27  Asian  ...       No   
15  Graduate school and higher      None  26  Asian  ...       No   
16                 High school  A garden  17  Asian  ...       No   
17                 High school     A pet  17  Asian  ...       No   
18                 High school      None  22  Asian  ...       No   
19                 High school      None  21  Asian  ...       No   

   Pointandclick  RPG Simulation Sports  Story-driven  Strategy  \
0             No  Yes        Yes     No            No       Yes   
1             No   No        Yes     No            No       Yes   
2             No  Yes         No     No            No       Yes   
3             No  Yes        Yes     No            No       Yes   
4             No  Yes        Yes     No            No       Yes   
5             No  Yes        Yes     No            No       Yes   
6             No  Yes        Yes     No            No       Yes   
7             No  Yes        Yes    Yes            No        No   
8             No  Yes         No     No            No       Yes   
9             No  Yes        Yes     No            No       Yes   
10            No  Yes        Yes     No            No        No   
11            No  Yes        Yes     No            No        No   
12            No  Yes        Yes     No            No        No  

In [233]:
survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 640 entries, 0 to 639
Columns: 110 entries, Gamer_ID to vrchat
dtypes: int64(77), object(33)
memory usage: 550.1+ KB
